### ISSUE TITLE
Building exiv2 libraries under Win 10 with MSVC 2017

### ISSUE THREAD

**Issue by tester0077** [Exiv2/exiv2#901]
After some time off, I was coming back to try and use libexiv under Windows and I see that much has changed.
Is this the proper place to get help with sorting out the process of building the libs in this environment? Assuming so:
I have tried to follow the instructions for windows from https://github.com/Exiv2/exiv2/blob/master/README-CONAN.md#1-1, but am stuck when I try to run my first conan build.
It seems I have an installed version of Python 3.7.1 and have succeeded in installing conan - version1.16.0, but that is as far as I have gotten.
The next step from the instructions are 
cd $EXIV2_ROOT
but, this obviously fails, because I have not defined EXIV2_ROOT and have no clue what it should be

**Comment by clanmills** [comment]
We have two projects in progress at the moment:

v0.28 'master' is a major refactoring of the code for the future 
v0.27 '0.27-maintenance' are quarterly security fixes and minor changes

The EXIV2_ROOT is the location of the code. For example: /home/user/github/exiv2/master

**Comment by tester0077** [comment]
Thank you, that got me further along the job.
I was able to run

- conan install .. --build missing  --profile msvc2017Release64
- cmake ..  -G "Visual Studio 15 2017 Win64"
- cmake --build . --config Release

Where I get
```
 pngimage.cpp
     Creating library D:/pkg/C++/MSVC2017/exiv2-master-0.27.1/exiv2/build/lib/Release/exiv2.lib and object D:/pkg/C++/M
  SVC2017/exiv2-master-0.27.1/exiv2/build/lib/Release/exiv2.exp
convert.obj : error LNK2019: unresolved external symbol libiconv_open referenced in function "bool __cdecl `anonymous n
amespace'::convertStringCharsetIconv(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<ch
ar> > &,char const *,char const *)" (?convertStringCharsetIconv@?A0x9858b8c1@@YA_NAEAV?$basic_string@DU?$char_traits@D@
std@@V?$allocator@D@2@@std@@PEBD1@Z) [D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]
convert.obj : error LNK2019: unresolved external symbol libiconv referenced in function "bool __cdecl `anonymous namesp
ace'::convertStringCharsetIconv(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >
 &,char const *,char const *)" (?convertStringCharsetIconv@?A0x9858b8c1@@YA_NAEAV?$basic_string@DU?$char_traits@D@std@@
V?$allocator@D@2@@std@@PEBD1@Z) [D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]
convert.obj : error LNK2019: unresolved external symbol libiconv_close referenced in function "bool __cdecl `anonymous
namespace'::convertStringCharsetIconv(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<c
har> > &,char const *,char const *)" (?convertStringCharsetIconv@?A0x9858b8c1@@YA_NAEAV?$basic_string@DU?$char_traits@D
@std@@V?$allocator@D@2@@std@@PEBD1@Z) [D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]
C:\gnuwin32\lib\libiconv.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64' [D:
\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]
D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\bin\exiv2.dll : fatal error LNK1120: 3 unresolved externals [D:\pkg
\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]
```
This warning has me puzzled:
**C:\gnuwin32\lib\libiconv.lib : warning LNK4272: library machine type 'x86' conflicts with target machine type 'x64' [D:
\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build\src\exiv2lib.vcxproj]**

The Gnuwin32 directory existed before I started on this, so the code for exiv2 found it, but I have no clue as to how to fix it. FWIW, there also is  a cygwin64 directory on the same drive, because at one time or other I had used both gnuwin32 & sygein64 as part of some other work

**Comment by clanmills** [comment]
Ah, yes.  I've seen conflicts like this before.

I use the script cmd64.cmd in README.md to start cmd.exe with a simplified path and other environment strings.  I do this to avoid collisions with cygwin and other stuff on the build machine.

The error you have reported looks correct.  You have a 32-bit build on libiconv.lib which CMake wants to link.  On my Windows-10 system, I don't have libiconv.lib and CMake reports:
```
Could NOT find Iconv (Missing Iconv_LIBRARY Iconv_INCLUDE_DIR)
```
The code to discover Iconv is in cmake/findDepenencies.cmake
```
1282 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/cmake $ grep -Hn find findDependencies.cmake 
findDependencies.cmake:4:# Check if the conan file exist to find the dependencies
findDependencies.cmake:11:find_package(Threads REQUIRED)
findDependencies.cmake:14:    find_package( ZLIB REQUIRED )
findDependencies.cmake:19:        find_package(CURL REQUIRED)
findDependencies.cmake:23:        find_package( SSH REQUIRED)
findDependencies.cmake:31:        find_package(EXPAT REQUIRED)
findDependencies.cmake:33:        find_package(XmpSdk REQUIRED)
findDependencies.cmake:39:    find_package(Intl REQUIRED)
findDependencies.cmake:42:find_package(Iconv)   <-------- Comment this off to never search for Iconv
findDependencies.cmake:49:    find_program(CCACHE_FOUND ccache)
1283 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance/cmake $ 
```

For cygwin builds, I have recommended (in README.md 5.4 Cygwin) that you download and build libiconv-1.15.tar.gz  I have not recommended this for Visual Studio, so I suspect I have never got iconv to work with Visual Studio builds.  The library iconv is used to do character set conversions and it is benign to omit this from Exiv2.

Thanks for working on this and **digging in** to make progress.  I'm happy to help you 1-to-1 on Skype (or any screen share tool).  I'm 100% confident that Exiv2 0.27-maintenance will build on our supported platforms.

**Comment by tester0077** [comment]
My apologies for the tardy reply; I was sorting out some details with the utilities I intend to use the Exiv2 library with, mainly to set things up to create a 64-bit version.
I have found findDependencies.cmake, and will see whether commenting out the section you mentioned will work. 
Overall, I expect that I will need  libiconv because I expect that there will be some Unicode encoded strings to be processed.
Basically, I guess I will have to familiarize myself with conan & refresh my memory on Cmake.
On this point, the conan profile seems to imply, at first reading, that it is intended to handle both 32 & 64 bit versions, because of:
[settings] arch=x86_64 arch_build=x86_64
Is that correct?
And, would it be possible to change the 'recipe' to build_type=Debug to have it create the debug version?

**Comment by tester0077** [comment]
What would it take to make a 32-bit version? - more generally useful under Windows

**Comment by clanmills** [comment]
Exiv2 builds in 32 bits on all platforms.  I think 32 bit died with XP, however I believe there are still lots of folks using XP!

For MSVC 2017:
```
conan install .. --build missing --profile msvc2017Release32
cmake .. -G "Visual Studio 15 2017"
cmake --build . --config Release
```
You will need the profile msvc2017Release32.  This is documented in README-CONAN.md  You can copy the profile from `<exiv2dir>\cmake\msvc_conan_profiles` to `%USERPATH%\.conan\profiles`

**Comment by tester0077** [comment]
Thank you for your comments and references. Some progress in compiling the 32 bit release and debug versions in the meantime, but I still have to sort out a few more things after I analyze the CMake files a bit more.
I have created a new build32 directory for my 32-bit builds, but for some of the resulting files, the output for one version ends up overwriting the output for the other.
Other than that, I am getting a better handle on how this build system is intended to work and with a bit more work, I am sure to get it all working for me.

**Comment by clanmills** [comment]
Thanks for the update.  I always use a clean directory for every "type" of build.  CMake really doesn't help you to build static/DLLs 64/32 Debug/Release in the same tree.  So do something like:
```
> cd <exiv2dir>
> mkdir staticbuild && pushed staticbuild
> conan ... 
> cmake ..
> cmake --build . --config Release
> popd
```
If you're familiar with Visual Studio, you'll be used to building different configurations into a single tree.  However, CMake is a Linux tool and has different ideas about how the world works.  You'll soon get used to it.  I'm a long term fan of Visual Studio (since 1994) and happy with the CMake/Conan tools for building Exiv2 with Visual Studio.

**Comment by tester0077** [comment]
I'll get used to it all - yet.  For me, MSVC is also the preferred environment.
But, before I forget again, I just want to mention, that it seemed I had to run the latest batch of commands in an Admin DOS box.
To help me remember and analyze the workflow better, I will start over from scratch ;-)

**Comment by clanmills** [comment]
Ah yes, I always use cmd.exe in "Administrator" mode.  And I always switch off the Account Protection stuff.  When you're developing code, Windows "security" gets in the way.

It's only when I replied last night that I realised I've been using Visual Studio for 25 years.  Doesn't seem possible that so much time has gone by.  It was called "Visual C++" at the start, then it became "Developer Studio" and now it's "Visual Studio".  Today, I develop Exiv2 in CLion on MacOS-X.  Netbeans gave 3 Free Licenses to Team Exiv2 for all their products (worth $1500/annum).  Very happy with CLion and it works really well on Linux and Windows (using GCC on MinGW).

I hope to release Exiv2 v0.27.2-RC1 today.  Going to Normandy on Friday to run in the D-Day Half Marathon on Sunday.  Exiv2 v0.27.2 is on track for release on 2019-06-30.

Anyway, best wishes with your project.  Thank You for working with Exiv2.

**Comment by tester0077** [comment]
I'll get used to it all - yet.  For me, MSVC is also the preferred environment.
But, before I forget again, I just want to mention, that it seemed I had to run the latest batch of commands in an Admin DOS box.
To help me remember and analyze the workflow better, I will start over from scratch ;-)

**Comment by tester0077** [comment]
Made some more progress and was able to compile a 32-bit version for both release and debug configurations by building in separate build dirs.
But I have run into an issue with a jpeg from my Canon Powershot. I'll try to attach a sample picture as well as the text output I got from exiv2 via:
D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build32Release\bin>exiv2 D:\Public\Pictures\160_1702\IMG_0004.JPG >test.txt
Note: looking at test.txt in Notepad++ shows a whole lot of NULLs at the end of the file
![IMG_0004](https://user-images.githubusercontent.com/9053628/59471266-791aa400-8def-11e9-997b-d8627498f7ce.JPG)

**Comment by D4N** [comment]
tester0077 <notifications@github.com> writes:

> Made some more progress and was able to compile a 32-bit version for both release and debug configurations by building in separate build dirs.

That's the way it should be done with CMake. While it supports in-source
builds, you really shouldn't perform them.

> But I have run into an issue with a jpeg from my Canon Powershot. I'll try to attach a sample picture as well as the text output I got from exiv2 via:
> D:\pkg\C++\MSVC2017\exiv2-master-0.27.1\exiv2\build32Release\bin>exiv2 D:\Public\Pictures\160_1702\IMG_0004.JPG >test.txt
> Note: looking at test.txt in Notepad++ shows a whole lot of NULLs at the end of the file
> ![IMG_0004](https://user-images.githubusercontent.com/9053628/59471266-791aa400-8def-11e9-997b-d8627498f7ce.JPG)
>

I get that too with exiv2 built from master. Have you considered that
the Exif comment of this image might actually contain NULLs only? I've
checked the JPEG and exiv2 claims to find the comment at 0x142e, where
there's a lot of NULLs.

**Comment by tester0077** [comment]
Thank you for your analysis and information.
My main interest is in the exiv2 library for Windows under MSVC 2017 (preferably statically linked) and the only version that is currently available pre-built 64-bit release DLL version, I had to dig into the build process. As it is, I am not very familiar with CMake, though I have used it in the past, but in a Ubuntu environment.
The need to build in a separate build environment is a good idea, though there are some questions in my mind as to why I need separate build directories for debug & release versions.
Let me explain my questions on that issue: the export libraries needed to link t exiv2.lib end up in 
...... exiv2\build32Debug\lib\Debug  but al of the rest, including the DLL, end up in
......exiv2\build32Debug\bin
As it is, in order to be able to debug an application using the exiv2 lib, I also need the debug version of the DLL( which is much larger than the release version), why that file cannot be also be placed in the ..\Debug directory is the puzzle I am trying to unravel. In my work flow, I do go back and forth between debug and release version and I would like to have both directories available at all times.
...........
As for the nulls, they may well be a part of the file, but the other utility (Metadata++ based on Phil Harvey's Exiftool) I have tried to crosscheck the Exxiv2.exe output seems to be able to handle them.
At this point I am not very familiar with the standard these images are expected to conform to, but exifprint shows a similar pattern
Exif.Photo.UserComment                       0x9286 Undefined 264                              

but at least it recognizes that there seems to be an issue.
FWIW, this file came directly from the camera, which apparently means that Canon camera software is at the bottom of it all.

**Comment by clanmills** [comment]
I'm off to Normandy to run in the D-Day Marathon on Sunday.  I'll look at this next week when I get home.  Can you provide me with you test file and command-line arguments, please.

**Comment by tester0077** [comment]
The test file is the one attached above, showing the battered mouse.
Command line was simply either exiv2 or exifprint followed by the file path and name.
If I get further along, I will try to investigate more inside the exiv2 library and let you know.

**Comment by clanmills** [comment]
Thanks for the update.

**Comment by tester0077** [comment]
I don't know much about the 'insides' of Exiv2 code, but, FWIW,  both Exiv2 & exifprint end up at a place in tiffcomposite_int.cpp at line 1843++
where
    // free functions
    TypeId toTypeId(TiffType tiffType, uint16_t tag, IfdId group)
    {
        TypeId ti = TypeId(tiffType); <<<<< ti = undefined 
        // On the fly type conversion for Exif.Photo.UserComment
        if (tag == 0x9286 && group == exifId && ti == undefined) {
            ti = comment;
        }
        .....
    }

**Comment by clanmills** [comment]
I've deferred this to v0.27.3 which is scheduled for 2019-09-30.  The priority at the moment is to release v0.27.2, have a vacation and deal with various family matters in July.

**Comment by piponazo** [comment]
@tester0077 Would you mind to change the title of the issue to describe your current issue? After reading the whole thread of messages I understood that you managed to build exiv2, right ?

**Comment by clanmills** [comment]
@tester0077 Can we close this?

From your comments in #976, I believe you've been able to build successfully with Visual Studio + Conan + CMake + README-CONAN.md

**Comment by tester0077** [comment]
Yes

**Comment by clanmills** [comment]
@tester0077 Good News!  Exif.Photo.UserComment has received major attention for v0.27.3.  The whole business came to the surface while working on #1046.  Phil was very helpful and explained it to me.  I've updated the man page to explain how to use charset=Unicode (and other encodings).

You hit the hot spot in the code.  `toTypeId()` has been modified to detect `comment`.
```
    TypeId toTypeId(TiffType tiffType, uint16_t tag, IfdId group)
    {
        TypeId ti = TypeId(tiffType);
        // On the fly type conversion for Exif.Photo.UserComment, Exif.GPSProcessingMethod, GPSAreaInformation
        if ( const TagInfo* pTag = ti == undefined ? findTagInfo(tag,group) : NULL ) {
            if ( pTag->typeId_ == comment ) {
                ti = comment;
            }
        }
        // ...
        return ti;
    }
```

You are quite correct in your observation of nul bytes in the output stream.  This resulted in issues with our test suite on UNIX (also new for v0.27.3).  The exiv2 command-line program (and other test programs) have been modified to output using `  std::out << .....toString()` instead of `  std::out << .....value()`.  So you should never encounter binary in the output.

You've asked me in the past about Unicode.  Like most native english speakers, I am very challenged in dealing with anything other that 7 bit ascii!  The subject of using different encodings in IPTC was raised this weekend in #1203

I would very much value your help with testing v0.27.3.  You can download it from https://pre-release.exiv2.org.  The current pre-release v0.27.3 RC1 has the updated man page concerning UserComment.  The changes to eliminate binary output will be in v0.27.3 RC2 which is scheduled for 2020-05-31.
